staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go

```diff
@@ -130,6 +132,21 @@ func validateStructuralInvariants(s *Structural, lvl level, fldPath *field.Path)
 	}
 
 	// restrict metadata schemas to name and generateName only
+	if kind, found := s.Properties["kind"]; found && checkMetadata {
+		if kind.Type != "string" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("properties").Key("kind").Child("type"), kind.Type, "must be string"))
+		}
+	}
+	if apiVersion, found := s.Properties["apiVersion"]; found && checkMetadata {
+		if apiVersion.Type != "string" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("properties").Key("apiVersion").Child("type"), apiVersion.Type, "must be string"))
+		}
+	}
+	if metadata, found := s.Properties["metadata"]; found && checkMetadata {
+		if len(metadata.Type) > 0 && metadata.Type != "object" {
```

Comment:
why would we allow metadata type to be ""?

Reply:
That wouldn't be allowed as types must be complete anyway. Fixed.